src/Components/Components/src/Routing/Router.cs

```diff
@@ -155,12 +190,51 @@ private void Refresh(bool isNavigationIntercepted)
             }
         }
 
+        private async Task RunOnNavigateAsync(string path)
+        {
+            // If this router instance does not provide an OnNavigateAsync parameter
+            // then we render the component associated with the route as per usual.
+            if (!OnNavigateAsync.HasDelegate)
+            {
+                return;
+            }
+
+            // If we've already invoked a task and stored its CTS, then
+            // cancel the existing task.
+            _onNavigateCts?.Dispose();
+
+            // Create a new cancellation token source for this instance
+            _onNavigateCts = new CancellationTokenSource();
+            var navigateContext = new NavigationContext(path, _onNavigateCts.Token);
+            var task = OnNavigateAsync.InvokeAsync(navigateContext);
+
+            // Create a cancellation task based on the cancellation token
+            // associated with the current running task.
+            var cancellationTaskSource = new TaskCompletionSource();
+            navigateContext.CancellationToken.Register(() =>
+                cancellationTaskSource.SetResult());
+
+            // If the user provided a Navigating render fragment, then show it.
+            if (Navigating != null && task.Status != TaskStatus.RanToCompletion)
+            {
+                _renderHandle.Render(Navigating);
+            }
+
+            await Task.WhenAny(task, cancellationTaskSource.Task);
+        }
+
+        private async Task RunOnNavigateWithRefreshAsync(string path, bool isNavigationIntercepted)
+        {
+            await RunOnNavigateAsync(path);
+            Refresh(isNavigationIntercepted);
+        }
+
         private void OnLocationChanged(object sender, LocationChangedEventArgs args)
         {
             _locationAbsolute = args.Location;
             if (_renderHandle.IsInitialized && Routes != null)
             {
-                Refresh(args.IsNavigationIntercepted);
+                _ = RunOnNavigateWithRefreshAsync(NavigationManager.ToBaseRelativePath(_locationAbsolute), args.IsNavigationIntercepted);
```

Comment:
Is it ok to ignore this task?

Reply:
We have a separate workitem to surface exceptions in this task, we don't want to tackle it on the current PR

Reply:
Not really. We have an open issue to track how we capture unhandled expectations and dispatch them to the renderer. It's tracked because the change is orthogonal to what we have here.

#23763